modules/apigee_edge_teams/src/Form/AddTeamMembersForm.php

```diff
@@ -73,7 +82,8 @@ class AddTeamMembersForm extends FormBase {
   public function __construct(TeamMembershipManagerInterface $team_membership_manager, EntityTypeManagerInterface $entity_type_manager) {
     $this->teamMembershipManager = $team_membership_manager;
     $this->userStorage = $entity_type_manager->getStorage('user');
-    $this->teamEntityType = $entity_type_manager->getDefinition('team');
+    $this->teamRoleStorage = $entity_type_manager->getStorage('team_role');
```

Comment:
There is a problem with loading entity handlers during a service constructor.

The this loads entity type definitions and if there are any services used during `hook_entity_type_alter`, you can get a circular dependency error.

This effectively only happens on cache rebuild but it will prevent the container from being rebuilt.

Reply:
This is a common pattern in Drupal core as well, therefore, I believe that there is no problem with retrieving saving entity storage handlers from the entity type manager in the constructor.
So you either found [a Drupal core bug](https://www.drupal.org/project/issues/drupal?text=circular+reference+dependency&status=Open&priorities=All&categories=All&version=All&component=All) or you are somehow misusing the combinations of services and `hook_entity_type_alter()` hook implementations.
I'll check a [MINT module](https://github.com/apigee/apigee-m10n-drupal), let's move this conversation to another channel.

Reply:
I've seen this in plugins but never in a service constructor.

Reply:
This is a form constructor, this is not the first place where this "pattern" is being used. So if it compiles, not like the param converter plugins (see below) I think there is no problem with that. The `hook_entity_type_alter()` issue that you have described is an edge case that should be handled in place. Multiple projects suffers from circular reference errors, sometimes only using `\Drupal::service()` instead of DI solves the problem as I could see:
https://www.google.hu/search?q=drupal+circular+reference+detected+for+service

Reply:
Yes, there shouldn't be any issue with this form.  The service that is causing us to use a workaround is `\Drupal\apigee_edge_teams\TeamPermissionHandler::__construct()`.